tests/ssg_test_suite/common.py

```diff
@@ -233,29 +248,155 @@ def _make_file_root_owned(tarinfo):
     return tarinfo
 
 
-def create_tarball():
+def _rel_abs_path(current_path, base_path):
+    """
+    Return the value of the current path, relative to the base path, but
+    resolving paths absolutely first. This helps when walking a nested
+    directory structure and want to get the subtree relative to the original
+    path
+    """
+    tmp_path = os.path.abspath(current_path)
+    return os.path.relpath(current_path, base_path)
+
+
+def template_tests(product=None):
+    """
+    Create a temporary directory with test cases parsed via jinja using
+    product-specific context.
+    """
+    # Set up an empty temp directory
+    tmpdir = tempfile.mkdtemp()
+
+    # We want to remove the temporary directory on failure, but preserve
+    # it on success. Wrap in a try/except block and reraise the original
+    # exception after removing the temporary directory.
+    try:
+        # Load product's YAML file if present. This will allow us to parse
+        # tests in the context of the product we're executing under.
+        product_yaml = dict()
+        if product:
+            yaml_path = product_yaml_path(SSG_ROOT, product)
+            product_yaml = load_product_yaml(yaml_path)
+
+        # Below we could run into a DocumentationNotComplete error. However,
+        # because the test suite isn't executed in the context of a particular
+        # build (though, ideally it would be linked), we may not know exactly
+        # whether the top-level rule/profile we're testing is actually
+        # completed. Thus, forcibly set the required property to bypass this
+        # error.
+        product_yaml['cmake_build_type'] = 'Debug'
+
+        # Note that we're not exactly copying 1-for-1 the contents of the
+        # directory structure into the temporary one. Instead we want a
+        # flattened mapping with all rules in a single top-level directory
+        # and all tests immediately contained within it. That is:
+        #
+        # /group_a/rule_a/tests/something.pass.sh -> /rule_a/something.pass.sh
+        for dirpath, dirnames, _ in walk_through_benchmark_dirs(product):
```

Comment:
It would be great to extract contents of this for loop to a separate function or method.

Reply:
We can do that in #7211.